File: Swift/ExampleTests/common/FandCString.swift
public struct FandCString: CustomStringConvertible, Equatable {
    public let f: String
    public let c: String
    public let notes: String

    public init(f: String, c: String, notes: String) {
        self.f = f
        self.c = c
        self.notes = notes
    }

    public init(fromArray v: [String]) {
        self.f = v.count > 0 ? v[0] : ""
        self.c = v.count > 1 ? v[1] : ""
        self.notes = v.count > 2 ? v[2] : ""
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    public static func fromText(_ text: String) -> FandCString {
        let parts = Tokens.require(text, 3, "FandC")
        return FandCString(f: parts[0], c: parts[1], notes: parts[2])
    }

    public var description: String {
        return Tokens.token(f) + " " + Tokens.token(c) + " " + Tokens.token(notes)
    }

    public static let dncString = "?DNC?"

    public static func == (a: FandCString, b: FandCString) -> Bool {
        return (a.f == dncString || b.f == dncString || a.f == b.f)
            && (a.c == dncString || b.c == dncString || a.c == b.c)
            && (a.notes == dncString || b.notes == dncString || a.notes == b.notes)
    }
}

File: Python/ExampleTests/common/fand_c_string.py
from . import tokens
DNC_STRING = '?DNC?'


class FandCString:
    def __init__(self, f: str = '', c: str = '', notes: str = ''):
        self.f = f
        self.c = c
        self.notes = notes

    @classmethod
    def from_list(cls, values):
        v = list(values)
        return cls(
            v[0] if len(v) > 0 else '',
            v[1] if len(v) > 1 else '',
            v[2] if len(v) > 2 else ''
        )

    @classmethod
    def from_text(cls, text):
        """Builds from the text form, e.g. Money as '25 USD'."""
        parts = tokens.require(text, 3, 'FandC')
        return cls(
            parts[0],
            parts[1],
            parts[2]
        )

    def __str__(self):
        return (tokens.token(self.f) + ' ' +
                tokens.token(self.c) + ' ' +
                tokens.token(self.notes))

    def _key(self):
        return (self.f, self.c, self.notes)

    def __eq__(self, other):
        if not isinstance(other, FandCString):
            return NotImplemented
        return all(a == b or a == DNC_STRING or b == DNC_STRING
                   for a, b in zip(self._key(), other._key()))

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'FandCString({self})'
